k9mail/src/main/java/com/fsck/k9/autocrypt/AutocryptHeader.java

```diff
@@ -16,16 +21,84 @@
     static final String AUTOCRYPT_PARAM_PREFER_ENCRYPT = "prefer-encrypt";
     static final String AUTOCRYPT_PREFER_ENCRYPT_MUTUAL = "mutual";
 
+    private static final int HEADER_LINE_LENGTH = 76;
 
+
+    @NonNull
     final byte[] keyData;
+    @NonNull
     final String addr;
+    @NonNull
     final Map<String,String> parameters;
     final boolean isPreferEncryptMutual;
 
-    AutocryptHeader(Map<String, String> parameters, String addr, byte[] keyData, boolean isPreferEncryptMutual) {
+    AutocryptHeader(@NonNull Map<String, String> parameters, @NonNull String addr,
+            @NonNull byte[] keyData, boolean isPreferEncryptMutual) {
         this.parameters = parameters;
         this.addr = addr;
         this.keyData = keyData;
         this.isPreferEncryptMutual = isPreferEncryptMutual;
     }
+
+    String toRawHeaderString() {
+        if (!parameters.isEmpty()) {
+            throw new UnsupportedOperationException("arbitrary parameters not supported");
+        }
+
+        String autocryptHeaderString = AutocryptHeader.AUTOCRYPT_HEADER + ": ";
+        autocryptHeaderString += AutocryptHeader.AUTOCRYPT_PARAM_ADDR + "=" + addr + ";";
+        if (isPreferEncryptMutual) {
+            autocryptHeaderString += AutocryptHeader.AUTOCRYPT_PARAM_PREFER_ENCRYPT + "=" +
+                    AutocryptHeader.AUTOCRYPT_PREFER_ENCRYPT_MUTUAL + ";";
+        }
+        autocryptHeaderString += AutocryptHeader.AUTOCRYPT_PARAM_KEY_DATA + "=" + ByteString.of(keyData).base64();
+
+        StringBuilder headerLines = new StringBuilder();
+        int autocryptHeaderLength = autocryptHeaderString.length();
+        for (int i = 0; i < autocryptHeaderLength; i += HEADER_LINE_LENGTH) {
```

Comment:
Header folding is only allowed in places where whitespace is allowed. Unfolding only gets rid of the newline, not the space at the start of the line.

Example:
```
Subject: Hello
 world
```

becomes "Hello world" when unfolded, not "Helloworld".

Reply:
hm yeah, so? This is the intuitive way to split lines, isn't it? granted not all b64 decoders ignore whitespace, so this might be worth a comment in the autocrypt spec, but implementation wise it seems sane enough to me. it does adhere to the unit tests we created, too

Reply:
I agree. If you make a comment in the spec that parsers can fold arbitrarily as needed when sending and should unfold and ignore all whitespace when recieving you can make the header as long as you like.

Thinking about it this header restriction is I expect why the PGP signature was distributed as a separate MIME part rather than X-PGP-Signature:

Reply:
They don't have to, it's enough if either the b64 decoders can deal with whitespace inbetween, or the autocrypt parsers do a `.replace(' ', null)` inbetween :)

Reply:
You assume you only need to insert line breaks in the `keydata` value. But really it could be anywhere if a parameter value is long enough.

Reply:
If mail libraries do automatic unfolding and only hand back a single line header body you'll end up with whitespace in random places, e.g. `addr=…;_noncritical=ThisIsAnImport antValue AndShouldNotContainASpace;keydata=…`

It's not an issue for us, so I'm fine with that folding rule. But it really should be documented in the spec.

Reply:
That's definitely something we need to consider better. Thanks, I'll try to bring it up on the mailing list.